### PR TITLE
Adding ContentFragment for Slack

### PR DESCRIPTION
**Pull request by lasryaric**
This is for https://github.com/orgs/dust-tt/projects/3?pane=issue&itemId=40635488

Sending context as ContentFragment from the Slack bot.

A few points:
- We start computing the ContentFragment as early as possible when we get the slack webhook because there is a lot of I/O needed to be done before answering.
- We skip all mentions / answers from/to the bot when building the content fragment.

### CODE REVIEW

File: connectors/src/connectors/slack/bot.ts
Comment (spolu) [inline connectors/src/connectors/slack/bot.ts:320]: `Slack thread messages` will be more appropriate I think as initial fragment and follow-up fragment?

Comment (lasryaric) [reply connectors/src/connectors/slack/bot.ts:320]: done thanks.

File: connectors/src/connectors/slack/bot.ts
Comment (spolu) [inline connectors/src/connectors/slack/bot.ts:533]: nit

**Comment by spolu** [comment]
Also before we can merge that we need an UI to display content fragment in Conversations :)

File: connectors/src/connectors/slack/bot.ts
Comment (lasryaric) [reply connectors/src/connectors/slack/bot.ts:305]: Sounds good.

**Comment by lasryaric** [comment]
> Also before we can merge that we need an UI to display content fragment in Conversations :)

Sounds good. I will take care of it.

**Comment by lasryaric** [comment]
@spolu you can start reviewing, there are a few things missing but should prevent from reviewing.
The main missing thing is adding a "provider" concept to ContentFragmentType because right now the Slack icon is hard coded.
We also need to do some basic testing on the anticipated features (summarize a thread and translate it to another language) and see if the prompt is correctly built. GPT keep answering me that it does not see the thread I am talking about when I ask it to make a summary. Looks like we need to be more explicit in the fact that this is a (Slack) thread.

**Comment by lasryaric** [comment]
And it looks like that. The green background color on the Slack avatar is because I need to publish a new version of sparkle to support ReactNode and backgroundColor on the avatar component (today it's either a react node or an http URL).
Screen Shot 2023-10-10 at 18 45 17

**Comment by spolu** [comment]
I believe this is very confusing to have something that looks like a message. You want something that is distinctively different from a message. Something close to a Chip should do the work pretty well.

**Comment by spolu** [comment]
Do you strongly feel like the version you propose here is the best we can use?

**Comment by lasryaric** [comment]
> Do you strongly feel like the version you propose here is the best we can use?

Not necessarily, this idea was coming from @philipperolet and I personally love it, I think it makes a lot of sense, it's a message "coming from slack".

But we can definitely try something else.

**Comment by lasryaric** [comment]
@fontanierh also found it confusing so happy to try something else.

**Comment by spolu** [comment]
But that is not a message right?

**Comment by spolu** [comment]
So mimicking a message is very likely to confuse the user no?

**Comment by spolu** [comment]
Also how would that work for document insertion?

**Comment by philipperolet** [comment]
Jumping in; the point not to confuse it with agent / user message is valid. An option we had discussed was to give the CF a different background color that assistant / user messages -- also something representing clearly the fact that it's a CF in the left column in lieu of an Avatar (e.g. Content specific picto, + words content fragment)

However, a consistent look and feel is also important so having the CF's "content" ( in the sense "info that we wish to show the user about this CF" so not necessarily the content of the fragement) aligned with the rest of the content, with the same style, felt right and is what led to suggest the use of a conversationmessage

**Comment by philipperolet** [comment]
At this point I would call @Duncid to the rescue with the :tophat:  :smile:

**Comment by spolu** [comment]
@Duncid suggested to “throw a box” which really is a simple chip component ImHO

**Comment by philipperolet** [comment]
A simple chip might not be very noticeable by the user (already a few chips in the agent messages), the CF might deserve a bit more. And there's the question of how it would be positioned to keep everything good-looking

@lasryaric had experimented with a box like the citations -- as ed suggested, but it didn't look sparkly

=> I can jump in with @lasryaric and make 2 propositions this morning (I have something in mind) and if they are voted down we can switch to anything else you/ed see fit?

**Comment by spolu** [comment]
A content fragment is a fragment of content. It can be a document injected in the conversation by the user, it can be the text of an opened tab added by the user, it can be some context from slack, it could be an automated summary of the conversation generated for very long conversations when necessary it could be notion page injected in context by some magic action on notion (eg @dust conversation:converstaionId as a comment) it could be a reply to an email that was sent from the conversation in the future.

Let's start from something simple that tells what it is and that is future proof to document injection from the input bar. It's a bunch of content that was added to the conversation. A message is a piece of text added to the conversation by a specific user or assistant. Slack is not a user. And we do not want to show the slack content in conversation that would be distasteful wouldn't it?

**Comment by spolu** [comment]
Also I strongly feel it has to be visually distinctive from a citation as they don't play at all the same role. A citation is a link to external content. A content fragment is external content injected in conversation.

**Comment by philipperolet** [comment]
Ok so the direction I was experimenting on below -- distinct from user & agent messages while a clear part of the conversation context
From the look of the above messages I gather this will not excite :sweat_smile: 
image

**Comment by philipperolet** [comment]
of course would require some tuning on the bg color & perfecting the left column but this is the idea

**Comment by philipperolet** [comment]
My questions about the chip were 1. noticeable enough? and 2. how exactly should it be placed to integrate seamlessly in the conversation and look good?

=> That being said, happy to commit to the chip or any other solution and assist @lasryaric to ship if he so wishes

**Comment by Duncid** [comment]
ideally (We have a user associated), I would do:
image

In the meantime:
image

**Comment by Duncid** [comment]
image

**Comment by Duncid** [comment]
image

**Comment by lasryaric** [comment]
Since the use case is Slack only for now and we are not even going to display the content of the content fragment, I think it does not matter too much what UI solution we are going with for now.
I will try it with a chip for now just because we need to make a decision (and I already tried citation and conversation), and we can revisit later when we have more concrete use case coming up.

**Comment by lasryaric** [comment]
@philipperolet @Duncid @spolu what do you think about this one?
Screen Shot 2023-10-11 at 12 20 41
(not the best example since the LLM did not get which thread I was talking about but I

**Comment by lasryaric** [comment]
And here it is when it's not Slack:
Screen Shot 2023-10-11 at 13 48 06

**Comment by spolu** [comment]
@Duncid You don't want to have the `@dust message` in the content fragment since it will be exactly the user message right after.

Good idea to have the channel in there.

@lasryaric let's use the design otherwise? I'm somewhat happy with a Chip as well. But def not in green. What do you want to do?

File: front/components/assistant/conversation/ContentFragment.tsx
Comment (spolu) [inline front/components/assistant/conversation/ContentFragment.tsx:15]: typo

**Comment by lasryaric** [comment]
I personally think the chip looks better than the citation.

File: front/components/assistant/conversation/ContentFragment.tsx
Comment (lasryaric) [inline front/components/assistant/conversation/ContentFragment.tsx:15]: I think tackling the issue of "source" of message definitely belongs to another PR and might even need a little design doc. So for now we are going with just checking the URL here, it works pretty well.

File: front/lib/models/assistant/conversation.ts
Comment (spolu) [inline front/lib/models/assistant/conversation.ts:380]: I thiiink we want to add a type as well. We can start with type "slack_thread_content" we'll add other content fragment types in the future this will avoid the switch on the slack.comm url which is not super future proof

File: front/components/assistant/conversation/ContentFragment.tsx
Comment (spolu) [reply front/components/assistant/conversation/ContentFragment.tsx:15]: Just commented. Let's add a type to content fragment :+1:

**Comment by lasryaric** [comment]
And I think the day we add support for uploading documents directly on the webapp this is going to change anyway, so might not worth spending too much time on it right now. Except if you think it's confusing or look bad.

**Comment by spolu** [comment]
@Duncid Also it will generally be a UserMessage after the content fragment which might impact your design as the background of the next message won't be white but the same light grey

**Comment by spolu** [comment]
The green won't fly with product I think. I think we're all aligned on the content and general shape of it, now we need to converge on a final design.

File: front/lib/models/assistant/conversation.ts
Comment (lasryaric) [reply front/lib/models/assistant/conversation.ts:380]: done

**Comment by lasryaric** [comment]
Happy to go with the citation design if that makes us save some time. I will do that.

**Comment by Duncid** [comment]
FYI
https://github.com/dust-tt/dust/pull/2056

**Review by spolu — COMMENTED** [review]
Left a few comments will make another pass on the PR today

**Comment by spolu** [comment]
Can you share a final screenshot as well? :pray: :pray:

**Review by spolu — DISMISSED** [review]
Other than my previous comments PR LGTM assuming we don't need migration :+1:

**Comment by lasryaric** [comment]
thanks @spolu . Working on merging it tonight.

**Comment by lasryaric** [comment]
Here is what it looks like now:
Screen Shot 2023-10-11 at 22 54 40

**Comment by lasryaric** [comment]
ok I am ready on my side, need to deploy sparkle tomorrow and then front. Won't do that tonight, too risky.

**Review by spolu — DISMISSED** [review]

File: front/lib/models/assistant/conversation.ts
Comment (fontanierh) [reply front/lib/models/assistant/conversation.ts:388]: [Not in prod](https://metabase.dust.tt/question#eyJkYXRhc2V0X3F1ZXJ5Ijp7InR5cGUiOiJuYXRpdmUiLCJuYXRpdmUiOnsicXVlcnkiOiJzZWxlY3QgY291bnQoKikgZnJvbSBjb250ZW50X2ZyYWdtZW50cyIsInRlbXBsYXRlLXRhZ3MiOnt9fSwiZGF0YWJhc2UiOjR9LCJkaXNwbGF5IjoidGFibGUiLCJwYXJhbWV0ZXJzIjpbXSwidmlzdWFsaXphdGlvbl9zZXR0aW5ncyI6e319)

Comment (spolu) [reply front/lib/models/assistant/conversation.ts:388]: Awesome

**Pull request merged**
